storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerManager.java

```diff
@@ -139,6 +163,35 @@ public void close() throws IOException {
 
     public void addAssignmentsForPartitions(Set<TopicIdPartition> partitions) {
         consumerTask.addAssignmentsForPartitions(partitions);
+        waitTillMetaPartitionConsumptionCatchesUp();
```

Comment:
Good point @kamalcph ! Do you have any suggestion here?

Reply:
I am also unclear as to what we should do here, I mean we could return a `CompletableFuture<Void>` but this would involve changing the interface?

Reply:
One approach we can take is to wait for metadata partition to catch up asynchronously. All the calls to read the remote log will hit the [RemotePartitionMetadataStore#getRemoteLogMetadataCache](https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/RemotePartitionMetadataStore.java#L147), there we can ensure that the partition is initialised via a latch. 

The [ConsumerTask](https://github.com/apache/kafka/blob/trunk/storage/src/main/java/org/apache/kafka/server/log/remote/metadata/storage/ConsumerTask.java) can keep a marker of the latest offset of metadata-partition when user-partitions are assigned to them and can mark that user-partition as initialised in the cache once it read/re-read messages up-to that point. 

We can have a latch with timeout of 2 mins to avoid holding the caller thread for a long amount of time and throw a retry-able error.

Reply:
That also works! We have to look around for all the callers:

1. Replica fetcher thread will retry the request after the `replica.fetch.backoff.ms` once the partition is marked as error/failed.
2. Consumers may resend the FETCH request as soon as it fails without the request quota. 

We can also think about adding one more API in [RemoteLogMetadataManager](https://github.com/apache/kafka/blob/trunk/storage/api/src/main/java/org/apache/kafka/server/log/remote/storage/RemoteLogMetadataManager.java) to ensure that the partition is initialised before copying the segments (or) when the replica thread is being initialised. 

```
boolean isInitialized(TopicIdPartition topicIdPartition) throws RemoteStorageException;
```